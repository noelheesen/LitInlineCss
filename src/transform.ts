import * as walk from 'acorn-walk'
import MagicString from 'magic-string'
import inlineStylesheet from './mutators/inline-stylesheet'
import removeImport from './mutators/remove-import'
import replaceReferences from './mutators/replace-references'
import createAST from './util/create-ast'
import findLitElementImport from './util/find-lit-element-import'
import immutable from './util/immutable'
import loadCSS from './load-css'
import processCSS from './process-css'
import type { DynamicAcornNode, LitInlineCSSResult, CSSResult } from './types'

/**
 * core transformer that loads all CSS imports and inlines them.
 * @param context transformer context
 * @param input raw TS/JS input
 */
const transform = async (
  source: string,
  absoluteSourcePath: string,
  options?: {
    moduleNameMappings?: Record<string, string>
    rootDir?: string
  }
): Promise<LitInlineCSSResult> => {
  const AST = createAST(source)
  const litImport = findLitElementImport(AST)
  const cssResultSet = new Set<Readonly<CSSResult>>()

  if (!litImport) {
    throw new Error('module does not import lit-element')
  }

  const cssImports = await loadCSS(AST, absoluteSourcePath, options)
  if (cssImports.size === 0) {
    return {
      js: source,
      css: cssResultSet,
    }
  }

  await Promise.all(
    Array.from(cssImports).map((cssImport) =>
      (async () => {
        const { absolutePath, importSource, raw, specifier } = cssImport
        const { css: processedCSS, classmap, dependencies } = await processCSS(
          raw,
          absolutePath
        )
        cssResultSet.add(
          immutable({
            absolutePath,
            css: processedCSS,
            classmap,
            dependencies,
            importSource,
            raw,
            specifier,
          })
        )
      })()
    )
  )

  const magicString = new MagicString(source)

  // add CSS specifier to lit-element import declaration
  if (litImport.namedSpecifier) {
    magicString.appendRight(litImport.namedSpecifier.end, `, css`)
  } else if (litImport.defaultSpecifier) {
    magicString.appendRight(litImport.defaultSpecifier.end, `, { css }`)
  }

  walk.full(AST, (node: DynamicAcornNode) => {
    for (const cssResult of cssResultSet) {
      for (const { should, mutate } of [
        inlineStylesheet,
        removeImport,
        replaceReferences,
      ]) {
        if (should(node, cssResult)) {
          mutate(node, magicString, cssResult, litImport)
        }
      }
    }
  })

  const js = magicString.toString()

  return {
    js,
    css: cssResultSet,
  }
}

export default transform
