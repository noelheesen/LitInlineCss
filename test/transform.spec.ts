import path from 'path'
import transform from '../src/transform'
import readFileSync from '../src/util/read-file-sync'

const fixture = (...partialPaths: string[]) =>
  path.resolve(__dirname, 'fixtures', ...partialPaths)

/**
 * given an input <source> it will
 * 1. inline all imported css wrapped in the lit-element css function as a template literal
 * 2. replace all references to the imported css object by it's values.
 */
describe('transformer', () => {
  it('throws when the source does not import lit-element', async () => {
    expect.assertions(1)

    const filename = 'does-not-import-lit-element.js'
    const absoluteSourcePath = fixture(filename)
    const source = readFileSync(absoluteSourcePath)

    await expect(transform(source, absoluteSourcePath)).rejects.toThrow()
  })

  it('appends named css import to lit-element with default import', async () => {
    expect.assertions(2)

    const filename = 'import-lit-default.js'
    const absoluteSourcePath = fixture(filename)
    const source = readFileSync(absoluteSourcePath)

    const { js } = await transform(source, absoluteSourcePath)

    expect(js).toMatch("import LIT, { css } from 'lit-element'")
    expect(js).toMatchSnapshot('import-lit-default')
  })

  it('appends css import to lit-element with named import', async () => {
    expect.assertions(2)

    const filename = 'import-lit-named.js'
    const absoluteSourcePath = fixture(filename)
    const source = readFileSync(absoluteSourcePath)

    const { js } = await transform(source, absoluteSourcePath)

    expect(js).toMatch("import { LitElement, html, css } from 'lit-element'")
    expect(js).toMatchSnapshot('import-lit-named')
  })

  it('appends named css import to lit-element with default and named import', async () => {
    expect.assertions(2)

    const filename = 'import-lit-default-named.js'
    const absoluteSourcePath = fixture(filename)
    const source = readFileSync(absoluteSourcePath)

    const { js } = await transform(source, absoluteSourcePath)

    expect(js).toMatch("import LIT, { html, css } from 'lit-element'")
    expect(js).toMatchSnapshot('import-lit-default-named')
  })

  it('does not append named css import when lit-element import is namespaced', async () => {
    expect.assertions(2)

    const filename = 'import-lit-namespaced.js'
    const absoluteSourcePath = fixture(filename)
    const source = readFileSync(absoluteSourcePath)

    const { js } = await transform(source, absoluteSourcePath)

    expect(js).toMatch("import * as LIT from 'lit-element'")
    expect(js).toMatchSnapshot('import-lit-namespaced')
  })

  describe('inlining stylesheet', () => {
    it('is a single assignment', async () => {
      expect.assertions(2)

      const filename = 'import-css-single.js'
      const absoluteSourcePath = fixture(filename)
      const source = readFileSync(absoluteSourcePath)

      const { js } = await transform(source, absoluteSourcePath)

      expect(js).toMatch(/\.styles\s=\scss`\.map/g)
      expect(js).toMatchSnapshot('inline-stylesheet-single-prototype')
    })

    it('is a member of an array assignment', async () => {
      expect.assertions(2)

      const filename = 'import-css-multi-prototype.js'
      const absoluteSourcePath = fixture(filename)
      const source = readFileSync(absoluteSourcePath)

      const { js } = await transform(source, absoluteSourcePath)

      expect(js).toMatch(/\.styles\s=\s\[css`(.|\s)+`, css`(.|\s)+`\]/g)
      expect(js).toMatchSnapshot('inline-stylesheet-multi-prototype')
    })
  })

  it('removes css import expression', async () => {
    expect.assertions(2)

    const filename = 'import-css-single.js'
    const absoluteSourcePath = fixture(filename)
    const source = readFileSync(absoluteSourcePath)

    const { js } = await transform(source, absoluteSourcePath)

    expect(js).not.toMatch(
      /import\s(.)+\sfrom\s('|")(.)+\.(css|sass|scss|styl(us)?|less)('|")/g
    )
    expect(js).toMatchSnapshot('remove-css-import-expression')
  })

  describe('replace references', () => {
    it('replaces references in object properties', async () => {
      expect.assertions(2)

      const filename = 'classmap-static-ref.js'
      const absoluteSourcePath = fixture(filename)
      const source = readFileSync(absoluteSourcePath)

      const { js } = await transform(source, absoluteSourcePath)

      expect(js).toMatch(
        /const classes\s=\s{(\n|\s|\r)+'a':\strue,(\n|\s|\r)+'a-b':\strue,(\n|\s|\r)+}/g
      )
      expect(js).toMatchSnapshot('classmap-static-ref')
    })

    it('in dynamic object specified in classMap', async () => {
      expect.assertions(2)

      const filename = 'classmap-element.js'
      const absoluteSourcePath = fixture(filename)
      const source = readFileSync(absoluteSourcePath)

      const { js } = await transform(source, absoluteSourcePath)

      expect(js).toMatch(
        /const classes\s=\sclassMap\({(\n|\s|\r)+'a':\strue,(\n|\s|\r)+'a-b':\strue,(\n|\s|\r)+}\)/g
      )
      expect(js).toMatchSnapshot('classmap-element')
    })

    it('in dynamic object inlined in classmap', async () => {
      expect.assertions(2)

      const filename = 'classmap-inline.js'
      const absoluteSourcePath = fixture(filename)
      const source = readFileSync(absoluteSourcePath)

      const { js } = await transform(source, absoluteSourcePath)

      expect(js).toMatch(
        /class=\${classMap\({(\n|\r|\s)+('|")a('|"):\strue,(\n|\r|\s)+('|")a-b('|"):\strue,(\n|\r|\s)+}\)}/g
      )
      expect(js).toMatchSnapshot('classmap-inline')
    })
  })
})
