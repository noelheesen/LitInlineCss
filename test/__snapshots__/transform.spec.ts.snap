// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer appends css import to lit-element with named import: import-lit-named 1`] = `
"import { LitElement, html, css } from 'lit-element'

class LitNamed extends LitElement {
  render() {
    return html\` <div>LitNamed</div> \`
  }
}

LitNamed.styles = css\`.map { display: block; }\`
customElements.define('lit-named', LitNamed)
"
`;

exports[`transformer appends named css import to lit-element with default and named import: import-lit-default-named 1`] = `
"import LIT, { html, css } from 'lit-element'

class LitDefaultNamed extends LIT.LitElement {
  render() {
    return html\` <div>LitDefaultNamed</div> \`
  }
}

LitDefaultNamed.styles = css\`.map { display: block; }\`
customElements.define('lit-default-named', LitDefaultNamed)
"
`;

exports[`transformer appends named css import to lit-element with default import: import-lit-default 1`] = `
"import LIT, { css } from 'lit-element'

class LitDefault extends LIT.LitElement {
  render() {
    return LIT.html\`
      <div>LitNamespaced</div>
    \`
  }
}

LitDefault.styles = css\`.map { display: block; }\`
customElements.define('lit-default', LitDefault)
"
`;

exports[`transformer does not append named css import when lit-element import is namespaced: import-lit-namespaced 1`] = `
"import * as LIT from 'lit-element'

class LitNamespaced extends LIT.LitElement {
  render() {
    return LIT.html\`
      <div>LitNamespaced</div>
    \`
  }
}

LitNamespaced.styles = LIT.css\`.map { display: block; }\`
customElements.define('lit-namespaced', LitNamespaced)
"
`;

exports[`transformer inlining stylesheet is a member of an array assignment: inline-stylesheet-multi-prototype 1`] = `
"import { LitElement, html, css } from 'lit-element'

class SingleComponent extends LitElement {
  render() {
    return html\`<div>Single Component</div>\`
  }
}

SingleComponent.styles = [css\`html,
body {
  margin: 0;
  padding: 0;
}
\`, css\`.map { display: block; }\`]

customElements.define('s-c', SingleComponent)
"
`;

exports[`transformer inlining stylesheet is a single assignment: inline-stylesheet-single-prototype 1`] = `
"import { LitElement, html, css } from 'lit-element'

class SingleComponent extends LitElement {
  render() {
    return html\`<div>Single Component</div>\`
  }
}

SingleComponent.styles = css\`.map { display: block; }\`

customElements.define('s-c', SingleComponent)
"
`;

exports[`transformer removes css import expression: remove-css-import-expression 1`] = `
"import { LitElement, html, css } from 'lit-element'

class SingleComponent extends LitElement {
  render() {
    return html\`<div>Single Component</div>\`
  }
}

SingleComponent.styles = css\`.map { display: block; }\`

customElements.define('s-c', SingleComponent)
"
`;

exports[`transformer replace references in dynamic object inlined in classmap: classmap-inline 1`] = `
"import { LitElement, html, css } from 'lit-element'
import { classMap } from 'lit-html/directives/class-map'

class LitClassmapInline extends LitElement {
  render() {
    return html\`
      <div
        class=\${classMap({
          'a': true,
          'a-b': true,
        })}
      >
        LitClassmapInline
      </div>
    \`
  }
}

customElements.define('lit-classmap-inline', LitClassmapInline)
"
`;

exports[`transformer replace references in dynamic object specified in classMap: classmap-element 1`] = `
"import { LitElement, html, css } from 'lit-element'
import { classMap } from 'lit-html/directives/class-map'

class LitClassmap extends LitElement {
  render() {
    const classes = classMap({
      'a': true,
      'a-b': true,
    })

    return html\`<div class=\\"\${classes}\\">LitClassmap</div>\`
  }
}

customElements.define('lit-classmap', LitClassmap)
"
`;

exports[`transformer replace references replaces references in object properties: classmap-static-ref 1`] = `
"import { LitElement, html, css } from 'lit-element'
import { classMap } from 'lit-html/directives/class-map'

const classes = {
  'a': true,
  'a-b': true,
}

class ClassmapStaticRef extends LitElement {
  render() {
    return html\` <div class=\${classMap(classes)}>ClassmapStaticRef</div> \`
  }
}

export default ClassmapStaticRef
"
`;
